src/System.Private.CoreLib/src/System/Threading/ThreadStore.cs

```diff
@@ -0,0 +1,57 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Runtime;
+
+namespace System.Threading
+{
+    internal static class ThreadStore
+    {
+        private static readonly ManualResetEvent _allDone = new ManualResetEvent(false);
+
+        // main thread started by runtime and we have no chance (?) to increment it elsewhere
+        private static int _foregroundRunningCount = 1;
+
+        internal static void IncrementRunningForeground() 
+        {
+            Interlocked.Increment(ref _foregroundRunningCount);
+        }
+
+        internal static void DecrementRunningForeground() 
+        {
+            if (Interlocked.Decrement(ref _foregroundRunningCount) == 0) 
+            {
+                _allDone.Set();
+            }
+        }
+
+        internal static void IncrementRunningForeground(Thread thread) 
+        {
+            if (!thread.IsBackground) 
+            {
+                IncrementRunningForeground();
+            }
+        }
+
+        internal static void DecrementRunningForeground(Thread thread) 
+        {
+            if (!thread.IsBackground) 
+            {
+                DecrementRunningForeground();
+            }
+        }
+
+        internal static void WaitForForegroundThreads() 
+        {
+            _allDone.WaitOne();
+        }
+
+
+        [RuntimeExport("RhpManagedShutdown")]
+        public static void RhpManagedShutdown()
```

Comment:
This is definitely a wrong place. Initially I've put it to [RuntimeExports.cs](https://github.com/dotnet/corert/blob/master/src/Runtime.Base/src/System/Runtime/RuntimeExports.cs) but I had to transitively reference ~200 files in `Runtime.Base.csproj` and `Test.CoreLib.csproj` which I guess is not good. Please show me the proper place or maybe there is a trick how to have it in `RuntimeExports` and not reference entire corelib.

Reply:
`StartupCodeHelpers.Shutdown` should be the best place to add it.